### DISCUSSION TITLE
How to manage the duration of an action

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#15977] [General]
Hi,
I am looking to control a motor (with an ESP32) for a short set duration, 500ms for example. I thought about this code but I think the motor stops for the time.sleep_ms, then resumes and so on. Which is not great
```
def forward(ms)
    for i in range (ms):
        motor.forward() #library dcmotor
        time.sleep_ms(1)    

forward(500)
```
I looked at the Timer function but the documentation states that this function uses an interrupt which by definition must be very short.
Maybe I should turn to asyncio?
Would you have a better idea to submit to me
Thanks in advance

**Comment by davefes** [comment]
Check out:
https://github.com/orgs/micropython/discussions/15969
I am doing something similar.

**Comment by peterhinch** [comment]
Using `asyncio` you might write something like:
```py
def forward(ms):
    asyncio.create_task(go_forward(ms))

async def go_foward(ms):
    motor.start()  # Start it running
    await asyncio.sleep_ms(ms)
    motor.stop()
```
The great advantage being that `forward()` returns immediately, enabling you to do other things while the motor starts and stops "in the background". See [asyncio tutorial](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md).

**Comment by bidrohini** [comment]
You can use the asyncio library to manage time-based tasks asynchronously, allowing you to control the motor without blocking other tasks. You can also use Millis-Based Non-Blocking Logic like this:
import time

def forward(ms):
    start_time = time.ticks_ms()
    motor.forward()  
    
    while time.ticks_diff(time.ticks_ms(), start_time) < ms:
        pass  
    
    motor.stop()  


forward(500)

**Comment by PatLub85** [comment]
Thanks for your suggestions.
Since I don't have much memory, the Timer solution works perfectly.
```
from machine import Timer
import dcmotor

tim1 = Timer(1)

def motor_off(timer):
    dc_motor.stop()

dc_motor.forward() # the motor runs for 2 seconds and stops
tim1.init(period=2000, mode=Timer.ONE_SHOT, callback=motor_off)
```
